flowauth/frontend/src/TokenBuilder.jsx

```diff
@@ -0,0 +1,185 @@
+import { withStyles } from "@material-ui/core/styles"
+import UserRoleList from "./UserRoleList";
+import { getDisabledState } from "rsuite/esm/CheckTreePicker/utils";
+import { Button, Dialog, DialogActions, DialogContentText, DialogTitle } from "@material-ui/core";
+import ScopedCssBaseline from "@material-ui/core/ScopedCssBaseline";
+import { scopes_with_roles } from "./util/util";
+import {createToken, getMyRolesOnServer} from "./util/api"
+import UserRolesPicker from "./UserRolesPicker";
+import SubmitButtons from "./SubmitButtons";
+import CompoundChecklist from "./TokenRolesPicker";
+import ScopeDetails from "./ScopeDetails";
+import TokenRolesPicker from "./TokenRolesPicker";
+
+const styles = (theme) => ({
+  root: {
+    ...theme.mixins.gutters(),
+    paddingTop: theme.spacing.unit * 2,
+    paddingBottom: theme.spacing.unit * 2
+  }
+})
+function TokenBuilder(props) {
+
+  const {activeServer, onClick} = props
+  
+  const [name, setName] = useState("")
+  const [nameHelperText, setNameHelperText] = useState("")
+  const [nameIsValid, setNameIsValid] = useState("")
+  const [roles, setRoleState] = useState([])
+  const [activeRoles, setActiveRoles] = useState([])
+  const [checked, setChecked] = useState([])
+  const [tokenErrorOpen, setTokenErrorOpen] = useState(false)
+  const [token, setToken] = useState("")
+  const [tokenError, setTokenError] = useState("")
+
+  // Run on initial load to get roles
+  useEffect(() => {
+    getMyRolesOnServer(activeServer)
+    .then(roles => setRoleState(roles),
+    (err) => console.log(err))
+  }, [])
+
+  // Run when setChecked is updated to make sure that this is reflected in activeScopes
+  useEffect(() => {
+    setActiveRoles(checked.map(i => roles[i]))
+    console.log("Active roles now:")
+    console.log(activeRoles)
+  }, [checked])
+
+  //Validates token name on change
+  useEffect(() => {
+    console.log("Name:" + name)
+    var letters = /^[A-Za-z0-9_]+$/;
+    if (name.match(letters)) {
+      setNameHelperText("");
+      setNameIsValid(true)
+    } else if (name.length === 0) {
+      setNameHelperText("Role name can not be blank.");
+      setNameIsValid(false)
+    } else {
+      setNameHelperText(
+        "Token name may only contain letters, numbers and underscores.",
+      )
+      setNameIsValid(false)
+    };
+  }, [name])
+
+  // Keeps track of which boxes are toggled
+  const handleToggle = (value) => () => {
+    const currentIndex = checked.indexOf(value);
+    const newChecked = [...checked]
+    console.log("box checked")
+    if (currentIndex === -1) {
+      newChecked.push(value)
+    } else {
+      newChecked.splice(currentIndex, 1)
+    }
+    setChecked(newChecked)
+  }
+
+  //Pops up a marquee containing the token for copy-and-paste or download
+  const requestToken = () => {
+    createToken(
+      name,
+      activeServer,
+      activeRoles.map(r => r.latest_token_expiry).reduce((prev,cur) => prev > cur ? prev : cur, new Date("1970")),
```

Comment:
This should be finding the _earliest_ `latest_token_expiry` across the active roles, not the latest (i.e. the latest expiry allowed by _all_ active roles, not the latest expiry allowed by _any_ active role).

In fact, since we're no longer letting the user set the expiry in the UI, I think this bit of logic should be pushed into the backend (i.e. remove the 'expiry' parameter from the "Create token" API route, and instead calculate the expiry in `add_token()` in the backend).

Reply:
In branch now